nodes/hoppers.lua

```diff
@@ -18,33 +18,45 @@ local function get_hopper_formspec(pos)
 	return formspec
 end
 
+-- unfortunately param2 overrides are needed for side hoppers even in the non-single-craftable-item case
+-- because they are literally *side* hoppers - their spouts point to the side rather than to the front, so
+-- the default item_place_node orientation code will not orient them pointing toward the selected surface.
 local hopper_on_place = function(itemstack, placer, pointed_thing, node_name)
-	local pos  = pointed_thing.under
-	local pos2 = pointed_thing.above
-	local x = pos.x - pos2.x
-	local z = pos.z - pos2.z
-
 	local returned_stack, success
-	-- unfortunately param2 overrides are needed for side hoppers even in the non-single-craftable-item case
-	-- because they are literally *side* hoppers - their spouts point to the side rather than to the front, so
-	-- the default item_place_node orientation code will not orient them pointing toward the selected surface.
-	if x == -1 and (hopper.config.single_craftable_item or node_name == "hopper:hopper_side") then
-		returned_stack, success = minetest.item_place_node(ItemStack("hopper:hopper_side"), placer, pointed_thing, 0)
-	elseif x == 1 and (hopper.config.single_craftable_item or node_name == "hopper:hopper_side") then
-		returned_stack, success = minetest.item_place_node(ItemStack("hopper:hopper_side"), placer, pointed_thing, 2)
-	elseif z == -1 and (hopper.config.single_craftable_item or node_name == "hopper:hopper_side")  then
-		returned_stack, success = minetest.item_place_node(ItemStack("hopper:hopper_side"), placer, pointed_thing, 3)
-	elseif z == 1 and (hopper.config.single_craftable_item or node_name == "hopper:hopper_side") then
-		returned_stack, success = minetest.item_place_node(ItemStack("hopper:hopper_side"), placer, pointed_thing, 1)
+
+	if not (hopper.config.single_craftable_item or node_name == "hopper:hopper_side") then
+		returned_stack, success = minetest.item_place_node(itemstack, placer, pointed_thing)
+		return returned_stack
+	end
+
+	local pointed_pos = pointed_thing.under
+	local hopper_pos  = pointed_thing.above
+
+	local param2
+	local pointed_registered = hopper.get_registered(minetest.get_node(pointed_pos).name)
+	if pointed_registered and pointed_registered["extra"]["set_hopper_param2"] then
```

Comment:
What's the origin of this `set_hopper_param2` function? I do not see any other uses in this mod.

Reply:
I wrote to api.md